### PR TITLE
ASoC: Reparenting fixes and suspend/resume errors propagation

### PR DESCRIPTION
**Pull request by crojewsk-intel**
The first two address the problem of ->runtime pointer shuffling when switching ownership of given BE e.g.: when during 2x FE <-> 1x BE scenario, the original FE which triggered the BE to be also created gets disconnected. First fix makes sure the ->private_data is not lost during said procedure. The second fix makes the pcm interface avoid re-parenting if a FE other than the original gets disconnected.

All other patches combined together allow for suspend() and resume() errors to be propagated up. Right now all the errors are squelched what limits the error-handling possibilities.

### CODE REVIEW

**Comment by crojewsk-intel** [comment]
The goal is to test the PR here so there is no regression on the SOF side. With that done, patchset will land on the mailing list for further review.

File: sound/soc/soc-pcm.c [hunk 1]
```diff
@@ -1264,6 +1264,9 @@ static void dpcm_be_reparent(struct snd_soc_pcm_runtime *fe,
 
 		fe_substream = snd_soc_dpcm_get_substream(dpcm->fe, stream);
 		be_substream->runtime = fe_substream->runtime;
+		/* Make sure private_data from the old FE is not lost. */
+		fe_substream = snd_soc_dpcm_get_substream(fe, stream);
+		be_substream->runtime->private_data = fe_substream->runtime->private_data;
```
Comment (plbossart) [inline sound/soc/soc-pcm.c]: FWIW. I looked at that and I removed the use of the private data in the BE in https://github.com/thesofproject/linux/pull/3908

I think it's completely wrong to share data from the FE in the BE. It should not be allowed, that's a design conceptual bug in DPCM.

In other words, the SOF parts don't care about this any longer.

Comment (crojewsk-intel) [reply sound/soc/soc-pcm.c]: I do agree. If there is no explicitly defined place where data between FE and BE can be exchanged, doing so implicitly through private_data sharing is not the correct way to handle things.

This problem, as most, is probably best addressed incrementally - fix the bug, then work on removing the flaw entirely.

**Comment by crojewsk-intel** [comment]
Added few debug prints as I could not navigate to the problem from available logs (dyndbg for snd_soc_core could be helpful, seems like logs here do not have it enabled). Patches show no issues in our tree for months so, perhaps some recent changes to the SOF driver combined with these few blow things up.

**Comment by crojewsk-intel** [comment]
The results seem to be hidden under 404 error this time. Any thoughts? Surely my patches did not crush Jenkins CI..

**Comment by plbossart** [comment]
> The results seem to be hidden under 404 error this time. Any thoughts? Surely my patches did not crush Jenkins CI..

There's an on-going migration from 01.org that is still going-on, so you have to look at the results from the Intel CI site.

**Comment by crojewsk-intel** [comment]
Thanks to Kai and Peter, managed to decipher the logs and get to the root of the problem. Results look green, except for one test marked as "FAIL (unconfirmed)".

Dropped one of the patches - `ASoC: pcm: Re-assign private date when reparenting`. The actual problem appears to be more general: `struct snd_pcm_substream` has no real `private_data`, not one available to a developer. What's present is occupied by the ASoC framework - substream's `private_data` stores pointer to an instance of `struct snd_soc_pcm_runtime`. So, a developer is left with `substream->runtime->private_data` which is lost during the reparenting process as the `->runtime` pointer is being re-assigned.

My suggestion? Let `snd_soc_pcm_runtime` extend `snd_pcm_runtime`, leaving the `->private_data` for developers.

**Comment by crojewsk-intel** [comment]
This PR fulfilled its purpose. Discussion regarding the reparenting issue - when HDAudio LINK stream is involved - is continued in #3987. Will send patches found here for upstream review.

**Comment by crojewsk-intel** [comment]
Resulting merge: https://lore.kernel.org/alsa-devel/20221116115550.1100398-1-cezary.rojewski@intel.com/.

**Pull request closed without merging**